### PR TITLE
Localization

### PR DESCRIPTION
**Pull request by tinygrox**
I localized more text

### CODE REVIEW

**Comment by tinygrox** [comment]
Oh the config files I also do some localized changes.
And how do I localize these places in _KerbalismConfig\System\Habitat.cfg_

I read the MM tutorial, but it made me even more confused.

image

**Comment by SirMortimer** [comment]
The days of the current habitat implementation are numbered, and that number isn't very high. Currently it's one of the more problematic code bases in Kerbalism and will be rewritten, hopefully soon.

**Comment by SirMortimer** [comment]
Looks good to me after a first glance. I'll look at it some more when I can

File: src/Kerbalism/UI/FileManager.cs [hunk 1]
```diff
@@ -131,12 +132,12 @@ static void Render_file(Panel p, uint partId, File file, Drive drive, bool short
 			if (file.transmitRate > 0.0)
 			{
 				if (file.size > 0.0)
-					exp_tooltip = Lib.Color(Lib.BuildString(exp_tooltip, "\nTransmitting at ", Lib.HumanReadableDataRate(file.transmitRate), " : <i>", Lib.HumanReadableCountdown(file.size / file.transmitRate), "</i>"), Lib.Kolor.Cyan);
+					exp_tooltip = Lib.Color(Lib.BuildString(exp_tooltip, "\n"+Localizer.Format("#KERBALISM_FILEMANAGER_TransmittingRate", Lib.HumanReadableDataRate(file.transmitRate)), " : <i>", Lib.HumanReadableCountdown(file.size / file.transmitRate), "</i>"), Lib.Kolor.Cyan);//Transmitting at <<1>>
```
Comment (gotmachine) [inline src/Kerbalism/UI/FileManager.cs]: don't use `+` and don't use the `<<1>>`, change to : 
`exp_tooltip = Lib.Color(Lib.BuildString(exp_tooltip, "\n", Localizer.Format("#KERBALISM_FILEMANAGER_TransmittingRate"), " ", Lib.HumanReadableDataRate(file.transmitRate), " : <i>", Lib.HumanReadableCountdown(file.size / file.transmitRate), "</i>"), Lib.Kolor.Cyan);`
with the string being `Transmitting at`

(same remark in other parts in this file)

**Comment by gotmachine** [comment]
Thanks for the contribution. There are a few things that require some editing before thsi can be merged:

- You deleted the KSP biniry zips, you need to revert that change.
- You localized a few things that aren't displayed and that shouldn't be changed in Telemetry.cs (see review)
- You should avoid to use the Localizer.Format() method that use the `<<x>>` syntax, it's awfully slow
- When string concatenation is involved using our `Lib.BuildString()` method, do not insert `+` operators inside it.  That method is there to avoid memory allocations and ultimately GC lag spikes by using a StringBuilder. 

I highlighted the places where those issues are the most problematic, but there are all over the place, so please check that in detail.

Either use only the `+` operator and remove the buildstring method or use a `Localizer.Format()` as a `Lib.buildString` argument. Performance wise, it's fine to use `+` concats in non-update places (for example for a one time action/message), but it's important to use the `Buildstring` method in anything that is updated every frame (update / fixedupdate or anything that appear in the kerbalism UI)

File: src/Kerbalism/UI/FileManager.cs [hunk 2]
```diff
@@ -156,13 +157,13 @@ static void Render_file(Panel p, uint partId, File file, Drive drive, bool short
...
-			p.AddRightIcon(Textures.toggle_red, "Delete the file", () =>
+			p.AddRightIcon(send ? Textures.send_cyan : Textures.send_black, Localizer.Format("#KERBALISM_FILEMANAGER_send"), () => { drive.Send(file.subjectData.Id, !send); });//"Flag the file for transmission to <b>DSN</b>"
+			p.AddRightIcon(Textures.toggle_red, Localizer.Format("#KERBALISM_FILEMANAGER_Delete"), () =>//"Delete the file"
 				{
-					Lib.Popup("Warning!",
-						Lib.BuildString("Do you really want to delete ", file.subjectData.FullTitle, "?"),
-				        new DialogGUIButton("Delete it", () => drive.Delete_file(file.subjectData)),
-						new DialogGUIButton("Keep it", () => { }));
+					Lib.Popup(Localizer.Format("#KERBALISM_FILEMANAGER_Warning_title"),//"Warning!"
+						Lib.BuildString(Localizer.Format("#KERBALISM_FILEMANAGER_DeleteConfirm", file.subjectData.FullTitle), "?"),//"Do you really want to delete <<1>>",
```
[the remark this answers is not in the document]
Comment (eggrobin) [reply src/Kerbalism/UI/FileManager.cs]: I don't think you can escape string interpolation for this one (it's a full sentence), compare:
```
#KERBALISM_FILEMANAGER_DeleteConfirm = Do you really want to delete <<1>>
```
with
```
#KERBALISM_FILEMANAGER_DeleteConfirm = 您真的要删除<<1>>吗
```

File: src/Kerbalism/UI/FileManager.cs (hunk 1 above)
Comment (eggrobin) [reply src/Kerbalism/UI/FileManager.cs]: Same here, it's a sentence:
Compare
```
#KERBALISM_FILEMANAGER_TransmittingRate = Transmitting at <<1>>
```
and
```
#KERBALISM_FILEMANAGER_TransmittingRate = 以<<1>>速率传输
```

You can probably escape string interpolation by turning this into a  `label: value` thing as you suggest, but you would have to change the string to a noun phrase, `Transmission rate:` or something along those lines.

Comment (gotmachine) [reply src/Kerbalism/UI/FileManager.cs]: Yup, well keep the localizer call, but at least change the `+` to a `,`

File: src/Kerbalism/UI/FileManager.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (tinygrox) [reply src/Kerbalism/UI/FileManager.cs]: @gotmachine  I'll use `Localizer.Format()` instead of `Lib.BuildString()` in this place, is that's ok?

**Comment by tinygrox** [comment]
Damn,it will take me a while.

File: src/Kerbalism/Modules/SolarPanelFixer.cs [hunk 3]
```diff
@@ -297,17 +298,17 @@ public void Update()
 					sb.Append(EcUIUnit);
 					if (analyticSunlight)
 					{
-						sb.Append(", analytic ");
+						sb.Append(", "+Localizer.Format("#KERBALISM_SolarPanelFixer_analytic") +" ");//analytic
```
[the remark this answers is not in the document]
Comment (tinygrox) [reply src/Kerbalism/Modules/SolarPanelFixer.cs]: Done

File: src/Kerbalism/UI/FileManager.cs [hunk 4]
```diff
@@ -75,8 +76,8 @@ public static void Fileman(this Panel p, Vessel v, bool short_strings = false)
 
 			if(filesCount > 0 || totalDataCapacity > 0)
 			{
-				var title = "DATA " + Lib.HumanReadableDataSize(usedDataCapacity);
-				if(!unlimitedData) title += Lib.BuildString(" (", Lib.HumanReadablePerc((totalDataCapacity - usedDataCapacity) / totalDataCapacity), " available)");
+				var title = Localizer.Format("#KERBALISM_FILEMANAGER_DataCapacity", Lib.HumanReadableDataSize(usedDataCapacity));//"DATA " +
```
[the remark this answers is not in the document]
Comment (tinygrox) [reply src/Kerbalism/UI/FileManager.cs]: As you wish

File: src/Kerbalism/UI/Monitor.cs [hunk 5]
```diff
@@ -408,7 +408,7 @@ void Problem_greenhouses(Vessel v, List<Greenhouse.Data> greenhouses, ref List<T
 				if (greenhouse.issue.Length > 0)
 				{
 					if (!icons.Contains(Textures.plant_yellow)) icons.Add(Textures.plant_yellow);
-					tooltips.Add(Lib.BuildString("Greenhouse: <b>", greenhouse.issue, "</b>"));
+					tooltips.Add(Lib.BuildString(Localizer.Format("#KERBALISM_Monitor_Greenhouse") +" <b>", greenhouse.issue, "</b>"));//"Greenhouse:"
```
[the remark this answers is not in the document]
Comment (tinygrox) [reply src/Kerbalism/UI/Monitor.cs]: Done

File: src/Kerbalism/UI/Planner/Planner.cs [hunk 6]
```diff
@@ -329,45 +330,45 @@ private static void AddSubPanelStress(Panel p)
...
 				() => { p.Prev(ref special_index, panel_special.Count); enforceUpdate = true; },
 				() => { p.Next(ref special_index, panel_special.Count); enforceUpdate = true; });
 
 			// render living space data
 			// generate details tooltips
 			string living_space_tooltip = Lib.BuildString
 			(
-				"volume per-capita:<b>\t", Lib.HumanReadableVolume(vessel_analyzer.volume / Math.Max(vessel_analyzer.crew_count, 1)), "</b>\n",
-				"ideal living space:<b>\t", Lib.HumanReadableVolume(PreferencesComfort.Instance.livingSpace), "</b>"
+				Localizer.Format("#KERBALISM_Planner_volumepercapita") +"<b>\t", Lib.HumanReadableVolume(vessel_analyzer.volume / Math.Max(vessel_analyzer.crew_count, 1)), "</b>\n",//"volume per-capita:
```
[the remark this answers is not in the document]
Comment (tinygrox) [reply src/Kerbalism/UI/Planner/Planner.cs]: Done

File: src/Kerbalism/UI/FileManager.cs (hunk 1 above)
Comment (tinygrox) [reply src/Kerbalism/UI/FileManager.cs]: Changed

**Comment by tinygrox** [comment]
I've changed all those "+" to "," in `Lib.BuildString()`(maybe)

**Comment by SirMortimer** [comment]
Tested it for a bit, looks good to me.

**Pull request merged**